### ISSUE TITLE
Consider making a TASTY backend

### ISSUE THREAD

**Issue by smarter** [twitter/rsc#21]
https://github.com/twitter/reasonable-scala/blob/master/docs/compiler.md currently says:
> Codegen
> We decided to postpone generation of executable code until we implement typechecking for a sizable subset of the language. Therefore, there is nothing to see in this section. Please come back later.

As I'm sure you're aware, going from the typechecker output to bytecode requires a huge amount of work. I wonder if you've considered outputting TASTY instead and using Dotty to emit bytecode ([and Javascript in the future](https://github.com/lampepfl/dotty/pull/3481), at least at first? This would allow you to validate the design of your typechecker without developing a full compilation pipeline, it would also bring some other benefits:
- Once you can unpickle TASTY, you can easily support separate compilation
- In my experience, TASTY is great for [IDE support](http://guillaume.martres.me/ide_paper.pdf)
- Two-way interoperability with Dotty becomes achievable (and maybe scalac interop in the future).

Currently, TASTY is informally specified in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala, and a proof-of-concept executable specification exists at https://github.com/DarkDimius/tasty-kaitai-spec

Of course you may already have completely different plans for the future, this is just an idea I wanted to share :)

**Comment by xeno-by** [comment]
I agree that it is hard to go from typed trees to executable code, and that’s why this functionality is currently not included in the near-future milestones of Rsc. Nonetheless, it’s good to discuss these things early.

It looks like we’re all thinking in the same direction - it seems beneficial to split the typechecker and the backend, connecting them by an intermediate representation (I’ll be referring to that as a typer IR). That is our current plan for emission of executable code in Rsc.

What currently unclear is the details of how to achieve this. Personally for me, there are two unknowns: 1) What should the typer IR look like? 2) How do we build a fast backend? Let’s go through these questions in order.

**What should the typer IR look like?**

First of all, this IR should be easy to emit and easy to consume. Apart from the backend, there are other tools that are interested in a typer IR: you mentioned IDEs, and in the Scalameta universe we have Scalafix, Metadoc, as well as [the nascent LSP implementation](https://github.com/scalameta/language-server) that are all backed by [SemanticDB](http://scalameta.org/tutorial/#SemanticDB). At the very minimum, there should be a lightweight library that can read and write the typer IR.

Secondly, since the only focus of Rsc is speed, we need an IR that is fast to read and write. We should probably have it friendly to incremental compilation, as raw compilation speed isn’t going to solve everything.

Finally, another mandatory feature is stability. We cannot change the IR on a whim - every change should result in a new version of the IR and should ideally provide an automated migration tool.

Our original plan here was to define a custom binary format optimized for performance and [use SemanticDB for tooling](https://github.com/twitter/reasonable-scala/issues/10), since there are already several high-profile tools for Scala that work with SemanticDB.

**How do we build a fast backend?**

One of the big themes that we’re going to explore in Rsc is parallelism. Rsc was carefully designed to enable parallel typechecking, and if we want to reap the benefits of that, in the long run we need to support parallel codegen as well.

Another theme that we should have in mind is runtime performance of the emitted executable code. If we are very fast at emitting code that is slow to run, that’s probably undesireable to our users. Therefore, we need the backend to be able to accommodate custom optimizations.

Towards that end, our original plan was to make use of the Scala Native compilation pipeline. It seems to have a parallel codegen as well as [a well-specified backend IR](http://scala-native.readthedocs.io/en/latest/contrib/nir.html). Whole-program optimizations supported by Scala Native also look interesting.

**Exploring the alternatives**

Rsc is an opensource project, and I think it is essential to openly evaluate alternatives, especially since both the typer IR and the fast backend aren’t in our immediate plans. I would be very interested to learn how the Dotty infrastructure fares with respect to the design requirements that we have in mind. I have several questions:

1) How does one emit TASTY? What are the expectations on tree shapes that the Dotty backend has about TASTY trees?
 
1) How does one read TASTY? Is there a library that tool authors can reference and use to load TASTY in their tools?

1) What is the performance story, in particular: how fast is TASTY emission and how does the binary size compare with classfile sizes?

1) Are there currently compatibility guarantees? If Dotty internals change, is TASTY going to change with them, too? What is the migration story?

1) Can the Dotty backend run in multiple threads?

1) How does the Dotty backend IR look like? Can you give an example of a couple optimizations that have been implemented on it?

**Comment by smarter** [comment]
Before answering your questions, I'd like to describe a bit the Dotty pipeline and the vocabulary we use to refer to each part so that we're all on the same page. The Dotty compiler is made up of many phases, the full list is at https://github.com/lampepfl/dotty/blob/e675e6b96a8cb52214e4813d4580a15ee146f136/compiler/src/dotty/tools/dotc/Compiler.scala#L42. It's a list of list phases where each sublist correspond to one AST traversal.
- The first phase (`FrontEnd`) is where the parsing/desugaring/naming/typing happen, we use the term **frontend** to refer to it. Its output is a typed AST and every phase after that takes a typed AST as input and output another typed AST.
- The last phase (`GenBCode`) is the one that actually emits bytecode, we use the term **backend** to refer to it. The Dotty backend is based on an early version of the Scala 2.12 backend, we would like to synchronize them eventually.
- Everything else is part of the **middle-end** for lack of a better word, we also call them **transform phases** since they live in the [transform/](https://github.com/lampepfl/dotty/tree/master/compiler/src/dotty/tools/dotc/transform) subdirectory and they transform trees.
- TASTY serialization happens in the `Pickler` phase. If rsc emitted TASTY, every phase after `Pickler` would have to be run by Dotty before emitting code.

Now, on to your questions!
> 1. How does one emit TASTY? What are the expectations on tree shapes that the Dotty backend has about TASTY trees?

TASTY is a binary format described in the links I gave in my original post above. Currently Dotty emits TASTY in a classfile attribute of `Foo.class` for every top-level class named `Foo`, but we also support reading and writing from `Foo.tasty` files that only contain the binary TASTY format. In practice you could probably copy-paste most of https://github.com/lampepfl/dotty/tree/master/compiler/src/dotty/tools/dotc/core/tasty and adapt the `Tree*.scala`, `Position*.scala` and `Name*.scala` files to use your data structures.

> 2. How does one read TASTY? Is there a library that tool authors can reference and use to load TASTY in their tools?

If you're willing to deserialize TASTY to the data structures used by the Dotty compiler, then you can use a Dotty compiler instance to do the job for you. For example @nicolasstucki is currently making a decompiler that way: https://github.com/lampepfl/dotty/pull/3485. This is also how the [Dotty IDE](http://dotty.epfl.ch/docs/usage/ide-support.html) works. If you want to use your own data structures you'll probably want to make your own deserializer anyway, but you could also transform the output of https://github.com/DarkDimius/tasty-kaitai-spec (in theory at least, I don't think anyone has tried doing that yet).

> 3. What is the performance story, in particular: how fast is TASTY emission [...]

We haven't made a serious effort to benchmark TASTY in isolation yet, but I can tell you that in our profiling, pickling takes a few % of running time. A lot of this time is spent doing local operations like computing deltas that should be really easy to run in parallel with the rest of the compiler without adding any locks, that's on my todo list to try :).

> [...] and how does the binary size compare with classfile sizes?

I just tried it on the `dotty-compiler` jar, here are my results:

|              | Without TASTY  | With TASTY (no positions) | With TASTY |
| ----------- | ------------- | - | ------------- |
| compressed .jar size | 8.6 MB | 9.5 MB | 12 MB |
| decompressed directory size  | 25 MB | 28 MB | 31 MB |

The "With TASTY (no positions)" column is like regular TASTY but without pickling the positions of tree nodes. Early on, we only stored fairly coarse positions in TASTY since they were only used to display error messages to the user, but since we started using TASTY in the Dotty IDE, we switched to pickling the position of every tree node. This has a size cost but allows us to implement go-to-definition/renaming/... easily. We're also considering adding scaladoc comments to TASTY trees to provide documentation support in the IDE (and the REPL!) and make documentation generation incremental.

> 4. Are there currently compatibility guarantees? If Dotty internals change, is TASTY going to change with them, too? What is the migration story?

The TASTY header contains a MajorVersion and a MinorVersion field: https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala#L24. MinorVersion bumps are backwards-compatible, MajorVersion bumps aren't. So far, we haven't tried very hard to keep compatibility since we weren't interoperating with anything else, but I think Scala.js .sjsir format has demonstrated that it's possible to evolve an AST representation while remaining binary-compatible for a long time.

> 5. Can the Dotty backend run in multiple threads?

By "Dotty backend" I assume you mean "Everything that runs after TASTY generation" so middle-end + backend. Currently, none of that is multi-threaded. There are efforts to parallelize the [scalac backend](https://github.com/scala/scala/pull/6124) that we would certainly want to integrate if they bear fruits. On our side, we are interested in trying to parallelize both our frontend and our middle-end. As you know, lazy completers for types makes this tricky, but that doesn't mean it's impossible ;).

> 6. How does the Dotty backend IR look like? Can you give an example of a couple optimizations that have been implemented on it?

Again, I'm assuming that we're talking about the middle-end and the backend here. The middle-end is typed ASTs just like in the frontend. We have a prototype of AST-based local optimizations in https://github.com/lampepfl/dotty/tree/master/compiler/src/dotty/tools/dotc/transform/localopt but it's not yet clear how useful this is. The backend doesn't do any optimization currently, but all the optimizations done by the Scala 2.12 backend could be applicable: https://github.com/scala/scala/tree/2.13.x/src/compiler/scala/tools/nsc/backend/jvm/opt
The frontend also implements `inline def` which guarantees method inlining (and prints an error if inlining is impossible), this can be used for manual optimizations.

**Comment by smarter** [comment]
I realize I forgot to answer part of the first question:
> What are the expectations on tree shapes that the Dotty backend has about TASTY trees?

The trees need to be normalized as described in the [header of `PostTyper`](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/PostTyper.scala#L19-L55).

**Comment by smarter** [comment]
Another thing worth noting is that TASTY does not store a type for every tree node, only definitions and leaf nodes get types. Types of intermediate expressions can be recovered using a [TypeAssigner](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala), the job done by the TypeAssigner is actually pretty close to the job currently being done by your typechecker (that is, it performs no adaptation/implicit search/type inference/..., it just assigns a type to a tree based on the type of its children).

**Comment by smarter** [comment]
Woops, my table of file sizes with TASTY was completely wrong because I was counting the TASTY section twice, I've just fixed it.

**Comment by xeno-by** [comment]
Thanks for the detailed answer! That was a lot of information, and I took a couple days to look into it in details. I thoroughly enjoyed the experience ❤️ 

All in all, I think that TASTY has a lot of potential, but right now it looks a bit rough for being a standard typer IR. I would like to suggest a few changes to the TASTY infrastructure that may be beneficial for its users. Some of these suggestions come from my previous experience of supervising a student project aimed at bringing TASTY support to Scalac. /cc @odersky @nicolasstucki.

**Suggestion 1: Specification**

I think that a more detailed specification of TASTY would be very welcome.

As you mentioned, there already is [a spec](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyFormat.scala), but it is quite informal. Without the knowledge of Dotty internals, I’m having a hard time understanding the meaning of various names, some trees (e.g. TERMREF*, INLINED, SHARED), types (e.g. TYPEREF*, RECtype), as well as some modifiers.

Moreover, from the current specification it is unclear how various language features are supposed to be expressed in TASTY. Back then when I was supervising the student project mentioned above, the only reliable way of figuring out TASTY representation of a given Scala snippet was running Dotty and inspecting its output, and it looks like it’s still the case these days. Having an implementation be the specification didn’t work well for the Scalac typechecker, so I think that potential TASTY users would appreciate more specificity here.

**Suggestion 2: Usability**

It would be great for usability if there was a standalone library that could take care of reading/writing TASTY.

You suggested copypasting [fragments of Dotty internals](https://github.com/lampepfl/dotty/tree/master/compiler/src/dotty/tools/dotc/core/tasty) to deal with this, but I’m a bit nervous about that. The suggested files are ~3.5kloc (not including the core data structures likes trees, types and the like). Moreover, as a prerequisite for TASTY being fully useful, you also mentioned [TypeAssigner](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/TypeAssigner.scala), which is itself >0.5kloc, not counting the dependencies on Dotty internal data structures that seem to span thousands if not tens of thousands lines of code. That’s a lot of code to copypaste and maintain.

**Suggestion 3: Compatibility**

Can you provide more information about compatibility guarantees? 

Given that Dotty hasn’t seen a stable release yet, how often do you plan to bump the MajorVersion number? If this number gets bumped, what will happen to alternative compilers that emit previous versions of TASTY? Will older TASTY payloads be still recognized and correctly processed by Dotty?

**Proposal**

I’d like to make a proposal that seems to address all the points that were raised above. 

How about we have a textual typer IR? Let’s define a new language called Core Scala that represents desugared Scala programs. Core Scala would look like normal Scala with potential minor changes necessary to accommodate typer output (e.g. to specify resolved names). In addition to its standard textual form, Core Scala can have implementation-defined binary representations that every compiler/toolchain generates however they want.

At a glance, Core Scala has the potential to solve a lot of issues:
1) Instead of painstakingly specifying AST nodes and how they represent Scala syntax, we simply piggyback on the existing syntax specification for Scala with potential minor changes to accommodate 3D syntax.
1) Usability-wise, reading and writing Core Scala won’t be very different from parsing and prettyprinting Scala. All compiler toolchains already have parsers and prettyprinters that work extremely fast (for example, on a case study project, the Rsc parser processes Scala code at ~1.3mloc per second).
1) Compatibility-wise, Core Scala seems to be easier to reason about than a binary format. We could mandate that if a Core Scala program parses, it should be supported by the toolchain, which is simpler to understand than tree shape preconditions required by the current implementation.

I am excited to explore Core Scala as Rsc develops. What do you guys think about this idea?

**Comment by smarter** [comment]
> Given that Dotty hasn’t seen a stable release yet, how often do you plan to bump the MajorVersion number?

I would say "as rarely as possible".

> If this number gets bumped, what will happen to alternative compilers that emit previous versions of TASTY? Will older TASTY payloads be still recognized and correctly processed by Dotty?

If we're serious about solving the binary compatibility problem of Scala the answer is yes. But of course, it's difficult for us to guarantee this until we've had more practical experience with it.

> 1. Instead of painstakingly specifying AST nodes and how they represent Scala syntax, we simply piggyback on the existing syntax specification for Scala with potential minor changes to accommodate 3D syntax.

What do you mean by "3D syntax"?

> 2. Usability-wise, reading and writing Core Scala won’t be very different from parsing and prettyprinting Scala. All compiler toolchains already have parsers and prettyprinters that work extremely fast (for example, on a case study project, the Rsc parser processes Scala code at ~1.3mloc per second).

Among other problems I can envision, t's going to be extremely difficult to guarantee that `Parser(PrettyPrinter(x)) == x`

> 3. Compatibility-wise, Core Scala seems to be easier to reason about than a binary format. We could mandate that if a Core Scala program parses, it should be supported by the toolchain, which is simpler to understand than tree shape preconditions required by the current implementation.

That means you'll need to encode these preconditions in the parser, that's also going to be difficult to do.

**Comment by odersky** [comment]
I think the proposal is not very reasonable (excuse the pun). You count 4k source-lines to adapt for a Tasty toolchain. How many freshly written sourcelines will you need for a source based version? How do you ensure you have accurate position information? (this required a huge effort in Tasty to get right. Like Guillaume says, how do you ensure that parsing is always the dual of pretty printing? 

I also disagree vehemently with the argument that a source-based scheme is easier to reason about. Syntax trees are arguably hard to reason about because they are quite large as ADTs. Going to source based simply sweeps that issue under a carpet. This might give a nice feeling of familiarity but it makes it really harder to reason about intermediate representations with confidence.

Take the issue of RecTypes that you mention. They express recursive references in a type such as

    C { type T; val x: this.T /*recursive!*/ }

They have a firm grounding in DOT where the theory of this is explained. If you go back to source and leave the user with arbitrary refinement types it might look more familiar but in the end a tool dealing with refinements will most likely forget that they can be recursive (and crash) or it will do the wrong thing, because doing the right thing is very hard! By using source you are just kicking the can down the road, so to speak. By contrast, going to a well-defined ADT requires some work for the encoder (which should be really just an adaptation of existing work) but then you have the confidence that you treatment of ASTs is exhaustive.

**Comment by odersky** [comment]
So I have a counter-proposal: Let's work together to make Tasty encoders for other compilers. We have a reasonable format that's already backed by a tool chain. Tasty is important because it is the distillation of a many years effort to find the simplest possible structures that describe full Scala. I think it would be very wasteful to throw away that effort and start from scratch.

**Comment by xeno-by** [comment]
> So I have a counter-proposal: Let's work together to make Tasty encoders for other compilers.

Clearly, TASTY has a lot of history behind it, and it would be great to have these experiences put to good use in the Scala community. To summarize my points from above, I think it would be beneficial to:
  1) Have a specification that explains things like RECtype and outlines how Scala language features map onto various TASTY constructs.
  1) Factor out the logic that deals with TASTY into a separate library like [ASM](http://asm.ow2.org), so that third-party applications don't have to depend on the entire Dotty compiler to work with it. In Scalameta, we've seen how separating tooling from compiler internals revitalized the Scala tooling ecosystem. I think it would be useful to leverage this experience in TASTY, too.
  1) Define a clear versioning policy for TASTY (e.g. no breaking changes except once a year or something).

**Comment by odersky** [comment]
Yes, I think these are excellent points to undertake. I would add:

  0. Clean up any things in Tasty that are Dotty-specific or backend specific. I believe we have some work to do here with regards to the different forms of names we expose. I'll try to get to that ASAP.

As to versioning, once products using Tasty are released I think we should have a very restrictive versioning policy. Once a year would be the absolute minimum, I would aim for more stability.

**Comment by mkeskells** [comment]
I think (hope) you meant to say "Once a year would be the absolute **max**imum"

**Comment by xeno-by** [comment]
There hasn't been much activity around this ticket in the last several months, so I'll be closing it for now. Let's revisit this once the points 0-3 from the comments above have been addressed.
